### PR TITLE
Closes #11163: Added separate method for client-side validation options in `yii\validators\Validator`

### LINKED ISSUE #11163: Add export of validation options without the js string
For example, now the clientValidateAttribute function  in  yii\validators\SomeValidator returns a string

``` php
public function clientValidateAttribute($model, $attribute, $view)
{
    $options = [
        'trueValue' => $this->trueValue,
        'falseValue' => $this->falseValue,
        'message' => Yii::$app->getI18n()->format($this->message, [
        'attribute' => $model->getAttributeLabel($attribute),
        'true' => $this->trueValue,
        'false' => $this->falseValue,
        ], Yii::$app->language),
    ];
    ///...
    ValidationAsset::register($view);

    return 'yii.validation.boolean(value, messages, ' . json_encode($options, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE) . ');';
}
```

This function can't be used without yii.validation.
Please, add new function or change this that it returned an array, like this

``` php
public function clientValidateAttribute($model, $attribute, $view)
 {
    $options = [
        'trueValue' => $this->trueValue,
        'falseValue' => $this->falseValue,
        'message' => Yii::$app->getI18n()->format($this->message, [
        'attribute' => $model->getAttributeLabel($attribute),
        'true' => $this->trueValue,
        'false' => $this->falseValue,
        ], Yii::$app->language),
    ];

    ///...

    return $options;
}
```

It is possible to make it so, for example

``` php
public function clientValidateOptions($model, $attribute)
{
    $options = [
        'trueValue' => $this->trueValue,
        'falseValue' => $this->falseValue,
        'message' => Yii::$app->getI18n()->format($this->message, [
        'attribute' => $model->getAttributeLabel($attribute),
        'true' => $this->trueValue,
        'false' => $this->falseValue,
        ], Yii::$app->language),
    ];
    // ... other setting options
    return $options;
}

public function clientValidateAttribute($model, $attribute, $view)
{
    $options = $this->clientValidateOptions($model, $attribute)
    ///...
    ValidationAsset::register($view);

    return 'yii.validation.boolean(value, messages, ' . json_encode($options, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE) . ');';
}
```

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #11163

### CODE REVIEW

**Review by samdark — APPROVED** [review]

**Comment by arogachev** [comment]
Should I add something else?

File: framework/validators/Validator.php [hunk 1]
```diff
@@ -361,6 +361,18 @@ public function clientValidateAttribute($model, $attribute, $view)
     }
 
     /**
+     * Returns the client-side validation options.
+     * @param \yii\base\Model $model the model being validated
```
Comment (SilverFire) [inline framework/validators/Validator.php]: I suggest to add the following lines:

```
This method is usually called from [[clientValidateAttribute()]].
You may override this method to modify options that will be passed to the client-side validation.
```

**Review by SilverFire — CHANGES_REQUESTED** [review]

**Comment by SilverFire** [comment]
Maybe we should PHPDoc for `clientValidateAttribute` as well? For example (spelling could be enhanced):

```
Calls [[getClientOptions()]] to generate options array for client-side validation.
...
@see getClientOptions()
```

**Comment by arogachev** [comment]
@SilverFire Yes, I thought about that. Will update it soon.

**Comment by arogachev** [comment]
Updated PHPDoc, please check.

**Review by SilverFire — APPROVED** [review]

**Review by dynasource — CHANGES_REQUESTED** [review]
is there a reason to have getClientOptions() to be public? From what I've seen in the code and read in the descriptions, it seems more logical to make it protected.  

For the rest this PR looks OK.

**Comment by SilverFire** [comment]
Makes sence

**Comment by arogachev** [comment]
@dynasource Yes, you are right. Fixed it.

**Review by dynasource — APPROVED** [review]

**Review by SilverFire — APPROVED** [review]

**Comment by SilverFire** [comment]
https://github.com/yiisoft/yii2/commit/cdaf5c86efe5a92159aabc44e36f42251d1c0d4c

**Comment by SilverFire** [comment]
Merged. Thank you for a great PR

**Comment by nkovacs** [comment]
Why is `getClientOptions()` protected?

It should be public so that a custom implementation of `ActiveField` can bypass `clientValidateAttribute` completely. That was the point of #11163.

(Side note: I was subscribed to #11163, but I didn't receive a notification of this pull request, because you only referenced the issue. I only got an email when the commit closed the issue, after the merge. Please comment on the issue next time, so that people watching it get a notification email.)

**Comment by cebe** [comment]
@nkovacs I have reopened #11163, lets discuss the details there.

**Pull request closed without merging**